### PR TITLE
KEP: Add kubectl plugin manager (krew)

### PR DESCRIPTION
**Pull request by lbb**

### CODE REVIEW

File: keps/sig-cli/0014-krew.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+## Summary
+
+krew is a package manager for the kubectl plugin system. krew introduces a way
+to handle kubectl plugin management and discovery.
+
+## Motivation
+
+The kubectl plugin system allows users to extend kubectl by placing extension
+executables in a plugin directory. This allows new subcommands and subcommand
+trees to be added in the kubectl CLI interface under `kubectl plugin <NAME>`.
```
Comment (duglin) [inline keps/sig-cli/0014-krew.md]: let's really see if we can remove the "plugin" word from the cmd line - that's a pretty big UX issue

Comment (duglin) [reply keps/sig-cli/0014-krew.md]: save the "plugin" word for managing plugins

Comment (lbb) [reply keps/sig-cli/0014-krew.md]: That makes sense. I think that issue is out of scope for this KEP and is related to https://github.com/kubernetes/community/blob/a6dcf8678b5c3febaacf079b28174b4a2cbd88e2/contributors/design-proposals/cli/kubectl-extension.md

File: keps/sig-cli/0014-krew.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+* Complete integration with various download sources like “hg://” or “samba://”
+* kubectl version compatibility management
+
+
+## Proposal
+
+#### Capabilities
+
+Note: This proposal has already been implemented in
+`github.com/GoogleContainerTools/krew` as a proof of concept.
```
Comment (soltysh) [inline keps/sig-cli/0014-krew.md]: Does not exist.

Comment (lbb) [reply keps/sig-cli/0014-krew.md]: The repository is now publicly accessible 😄 
Same URL: https://github.com/GoogleContainerTools/krew

**Comment by soltysh** [comment]
@lbb now that we can plugins up and ready can you update the current proposal and this can land?

**Comment by errordeveloper** [comment]
@soltysh are you referring to https://github.com/kubernetes/community/blob/master/keps/sig-cli/0024-kubectl-plugins.md? I'm finding it difficult to track the status of this KEP...

**Comment by ahmetb** [comment]
I'm tempted to hold off on this KEP until there is a kubectl release with the new plugin model shipped:

1. The plugin model change breaks krew and renders it unusable. We need to refactor krew and its plugin manifest to install plugins to somewhere in PATH.
2. We don't have enough plugins to understand if the manifest, installation system and the upgrade model is sufficient and satisfies users and packaging across multiple platforms (esp. Windows). We are working on getting more plugin devs to adopt the krew packaging format.

After we feel confident, we should come back to make progress on the KEP.

**Comment by soltysh** [comment]
Now that we have the plugin implementation ready as part of 1.12 release, this can continue.
/hold cancel

**Comment by ahmetb** [comment]
Small status update: we're still working on porting krew to the new plugin model introduced in 1.12. There are quite a lot of blocking changes, but we're close to releasing a krew v0.2 that's testable.

File: keps/sig-cli/0014-krew.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+  - sig-cli
+reviewers:
+  - TBD
+  - "@pwittroc"
+approvers:
+  - "@pwittroc"
+editor: lbb
+creation-date: 2018-07-03
+last-updated: yyyy-mm-dd
+status: provisional
```
Comment (mattfarina) [inline keps/sig-cli/0014-krew.md]: Given the discussion on the issue it sounds like this KEP is in the `implementable` status. See https://github.com/kubernetes/community/blob/master/keps/0001-kubernetes-enhancement-proposal-process.md#kep-workflow

File: keps/sig-cli/0014-krew.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+The goal is to move it into kubectl.
+
+Define a new meta config format called plugin manifest
+(stored in the index under: <plugin-name>.yaml)
+
+
+### Scenarios (standalone only)
+
+* User wants to install krew. They can run one `curl` command to install the
+  plugin manager, the plugin manager should then be able to function properly.
```
Comment (mattfarina) [inline keps/sig-cli/0014-krew.md]: Will the install scenario be the case or will it be apart of kubectl as earlier stated? Should this define the process for the proof of concept or what's going to be part of the toolchain?

File: keps/sig-cli/0014-krew.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
+---
+kep-number: 14
+title: Krew
```
Comment (mattfarina) [inline keps/sig-cli/0014-krew.md]: The name confused me when I first looked at this. I understand that krew is the name of the proof of concept. Should the title be more expressive of the purpose?

File: keps/sig-cli/0014-krew.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+* Developer wants to make his plugin discoverable by others.
+  They provide a link to the whole plugin as a compressed file under a public
+  URL. (Github provides any commit/tag as .tar.gz automatically.)
+  Afterwards the developer creates a PR against the index repository with a new
+  meta description file. This file contains the download url, download hash,
+  plugin version and description and tags. Meta descriptions are used by the
+  plugin manager to discover and manage plugins.
+* Developer wants to update their plugin and make those changes public.
+  They create a new PR against the index repository updating their plugin meta
+  description file.
```
Comment (mattfarina) [inline keps/sig-cli/0014-krew.md]: How do developers pin or otherwise deal with versions of plugins?

**Comment by mattfarina** [comment]
@soltysh Where is this at in terms of its lifecycle? How far is it from being merged where updates can be via other PRs? Is krew going to be moved under k8s control?

**Comment by justaugustus** [comment]
REMINDER: KEPs are moving to [k/enhancements](https://github.com/kubernetes/enhancements) on November 30. Please attempt to merge this KEP before then to signal consensus.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.

**Comment by justaugustus** [comment]
KEPs [have moved](https://github.com/kubernetes/enhancements/pull/638) to [k/enhancements](https://github.com/kubernetes/enhancements). 
This PR will be closed and any additional changes to this KEP should be submitted to k/enhancements.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.
/close

**Comment by dhiller** [comment]
@lbb I understand this PR should have been moved to k/enhancements. When will this happen?

**Pull request closed without merging**